### ISSUE TITLE
clippy::borrowed_box gives a a wrongly suggestion with std::mem::transmute

### ISSUE THREAD

**Issue by twhice** [rust-lang/rust-clippy#12602]
### Summary

wrong behavior on std::mem::transmute

### Lint Name

clippy::borrowed_box

### Reproducer

I tried this code:

```rust
fn main() {
    use std::mem::transmute;
    let boc = Box::new(1);

    unsafe {
        let reflect: Box<()> = transmute(boc);

        let reference1: &Box<i32> = transmute(&reflect);
        println!("{}", reference1);

        let reference2: &i32 = transmute(&reflect);
        println!("{}", reference2);
    }
}

```

clippy says:
```
warning: you seem to be trying to use `&Box<T>`. Consider using just `&T`
 --> src/main.rs:8:25
  |
8 |         let reference1: &Box<i32> = transmute(&reflect);
  |                         ^^^^^^^^^ help: try: `&i32`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#borrowed_box
  = note: `#[warn(clippy::borrowed_box)]` on by default
```

but its a UB! `&reflect` is a reference to `Box<()>` and actually a reference to `Box<i32>`

the suggestion turn `transmute::<&Box::<()>, &Box::<i32>>` into `transmute::<&Box::<()>, &i32`

### Version

```text
rustc 1.79.0-nightly (0824b300e 2024-03-24)
binary: rustc
commit-hash: 0824b300eb0dae5d9ed59719d3f2732016683d66
commit-date: 2024-03-24
host: x86_64-unknown-linux-gnu
release: 1.79.0-nightly
LLVM version: 18.1.2
```
